File: sdk/python/ai_os_sdk.py
"""
AI-OS Python SDK
Version: 1.0.0
Description: AI-OSプラットフォームとの統合を簡単にするためのPython SDK
"""

import os
import json
import time
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass
from datetime import datetime
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry


class AIOSError(Exception):
    """AI-OS API エラー"""
    
    def __init__(self, message: str, status_code: int, response: Optional[Dict] = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response


@dataclass
class Employee:
    """従業員データクラス"""
    id: str
    employee_code: str
    name: str
    email: str
    department: str
    position: str
    hire_date: str
    account_status: str
    metadata: Optional[Dict] = None


@dataclass
class TimeRecord:
    """勤怠記録データクラス"""
    id: str
    employee_id: str
    type: str
    timestamp: str
    location: Optional[Dict[str, float]] = None
    device: Optional[str] = None


@dataclass
class Payslip:
    """給与明細データクラス"""
    id: str
    employee_id: str
    year_month: str
    base_salary: float
    overtime_pay: float
    allowances: float
    gross_pay: float
    social_insurance: float
    income_tax: float
    total_deductions: float
    net_pay: float


@dataclass
class LeaveBalance:
    """休暇残高データクラス"""
    employee_id: str
    paid_leave: Dict[str, Any]
    sick_leave: Dict[str, Any]


@dataclass
class Expense:
    """経費データクラス"""
    id: str
    employee_id: str
    amount: float
    category: str
    date: str
    description: str
    status: str
    receipt_id: Optional[str] = None
    project_code: Optional[str] = None
    ai_recommendation: Optional[str] = None


class AIOSClient:
    """AI-OS APIクライアント"""
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://api.ai-os.com",
        version: str = "v1",
        timeout: int = 30,
        max_retries: int = 3,
        retry_delay: int = 1
    ):
        self.api_key = api_key or os.getenv("AIOS_API_KEY")
        if not self.api_key:
            raise ValueError("APIキーが必要です。引数または環境変数AIOS_API_KEYで設定してください。")
        
        self.base_url = base_url
        self.version = version
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        
        # セッション設定
        self.session = requests.Session()
        retry_strategy = Retry(
            total=max_retries,
            backoff_factor=retry_delay,
            status_forcelist=[429, 500, 502, 503, 504]
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # デフォルトヘッダー
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "X-SDK-Version": "1.0.0",
            "X-SDK-Language": "Python"
        })
        
        # APIエンドポイントの初期化
        self.employees = EmployeesAPI(self)
        self.time_records = TimeRecordsAPI(self)
        self.payroll = PayrollAPI(self)
        self.leaves = LeavesAPI(self)
        self.expenses = ExpensesAPI(self)
        self.reports = ReportsAPI(self)
    
    def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None,
        files: Optional[Dict] = None
    ) -> Union[Dict, List]:
        """HTTPリクエストを実行"""
        url = f"{self.base_url}/api/{self.version}{endpoint}"
        
        kwargs = {
            "timeout": self.timeout,
            "params": params
        }
        
        if files:
            kwargs["files"] = files
            # multipart/form-dataの場合はContent-Typeを削除
            headers = self.session.headers.copy()
            headers.pop("Content-Type", None)
            kwargs["headers"] = headers
        
        if data and method in ["POST", "PUT", "PATCH"]:
            kwargs["json"] = data
        
        try:
            response = self.session.request(method, url, **kwargs)
            response.raise_for_status()
            
            if response.content:
                return response.json()
            return {}
            
        except requests.exceptions.HTTPError as e:
            error_data = {}
            try:
                error_data = response.json()
            except:
                pass
            
            raise AIOSError(
                error_data.get("message", str(e)),
                response.status_code,
                error_data
            )
        except requests.exceptions.RequestException as e:
            raise AIOSError(str(e), 0)


class EmployeesAPI:
    """従業員管理API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def list(
        self,
        limit: int = 20,
        offset: int = 0,
        department: Optional[str] = None,
        status: Optional[str] = None
    ) -> Dict[str, Any]:
        """従業員一覧を取得"""
        params = {
            "limit": limit,
            "offset": offset
        }
        if department:
            params["department"] = department
        if status:
            params["status"] = status
        
        return self.client._request("GET", "/employees", params=params)
    
    def get(self, employee_id: str) -> Employee:
        """従業員詳細を取得"""
        data = self.client._request("GET", f"/employees/{employee_id}")
        return Employee(**data)
    
    def create(self, employee_data: Dict[str, Any]) -> Employee:
        """従業員を作成"""
        data = self.client._request("POST", "/employees", data=employee_data)
        return Employee(**data)
    
    def update(self, employee_id: str, employee_data: Dict[str, Any]) -> Employee:
        """従業員情報を更新"""
        data = self.client._request("PUT", f"/employees/{employee_id}", data=employee_data)
        return Employee(**data)
    
    def delete(self, employee_id: str) -> None:
        """従業員を削除"""
        self.client._request("DELETE", f"/employees/{employee_id}")


class TimeRecordsAPI:
    """勤怠管理API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def clock_in(self, employee_id: str) -> TimeRecord:
        """出勤打刻"""
        data = self.client._request(
            "POST",
            "/time-records/clock-in",
            data={"employee_id": employee_id}
        )
        return TimeRecord(**data)
    
    def clock_out(self, employee_id: str) -> TimeRecord:
        """退勤打刻"""
        data = self.client._request(
            "POST",
            "/time-records/clock-out",
            data={"employee_id": employee_id}
        )
        return TimeRecord(**data)
    
    def list(
        self,
        employee_id: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 20,
        offset: int = 0
    ) -> Dict[str, Any]:
        """勤怠記録一覧を取得"""
        params = {
            "limit": limit,
            "offset": offset
        }
        if employee_id:
            params["employee_id"] = employee_id
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        
        return self.client._request("GET", "/time-records", params=params)
    
    def get_monthly_summary(self, employee_id: str, year_month: str) -> Dict[str, Any]:
        """月次勤怠サマリーを取得"""
        return self.client._request(
            "GET",
            f"/time-records/summary/{employee_id}/{year_month}"
        )


class PayrollAPI:
    """給与管理API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def calculate(
        self,
        year_month: str,
        department_ids: Optional[List[str]] = None,
        employee_ids: Optional[List[str]] = None,
        include_bonus: bool = False,
        calculate_tax: bool = True,
        calculate_insurance: bool = True
    ) -> Dict[str, Any]:
        """給与計算を実行"""
        data = {
            "year_month": year_month,
            "include_bonus": include_bonus,
            "calculate_tax": calculate_tax,
            "calculate_insurance": calculate_insurance
        }
        if department_ids:
            data["department_ids"] = department_ids
        if employee_ids:
            data["employee_ids"] = employee_ids
        
        return self.client._request("POST", "/payroll/calculate", data=data)
    
    def get_payslip(self, employee_id: str, year_month: str) -> Payslip:
        """給与明細を取得"""
        data = self.client._request(
            "GET",
            f"/payroll/payslips/{employee_id}/{year_month}"
        )
        return Payslip(**data)
    
    def list_payslips(
        self,
        year_month: Optional[str] = None,
        department_id: Optional[str] = None,
        limit: int = 20,
        offset: int = 0
    ) -> Dict[str, Any]:
        """給与明細一覧を取得"""
        params = {
            "limit": limit,
            "offset": offset
        }
        if year_month:
            params["year_month"] = year_month
        if department_id:
            params["department_id"] = department_id
        
        return self.client._request("GET", "/payroll/payslips", params=params)
    
    def approve(self, batch_id: str) -> Dict[str, Any]:
        """給与計算結果を承認"""
        return self.client._request("POST", f"/payroll/batches/{batch_id}/approve")


class LeavesAPI:
    """休暇管理API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def request(
        self,
        employee_id: str,
        leave_type: str,
        start_date: str,
        end_date: str,
        reason: str
    ) -> Dict[str, Any]:
        """休暇申請"""
        data = {
            "employee_id": employee_id,
            "type": leave_type,
            "start_date": start_date,
            "end_date": end_date,
            "reason": reason
        }
        return self.client._request("POST", "/leaves/requests", data=data)
    
    def list_requests(
        self,
        employee_id: Optional[str] = None,
        status: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 20,
        offset: int = 0
    ) -> Dict[str, Any]:
        """休暇申請一覧を取得"""
        params = {
            "limit": limit,
            "offset": offset
        }
        if employee_id:
            params["employee_id"] = employee_id
        if status:
            params["status"] = status
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        
        return self.client._request("GET", "/leaves/requests", params=params)
    
    def get_balance(self, employee_id: str) -> LeaveBalance:
        """休暇残高を取得"""
        data = self.client._request("GET", f"/leaves/balance/{employee_id}")
        return LeaveBalance(**data)
    
    def approve(self, request_id: str) -> Dict[str, Any]:
        """休暇申請を承認"""
        return self.client._request("POST", f"/leaves/requests/{request_id}/approve")
    
    def reject(self, request_id: str, reason: str) -> Dict[str, Any]:
        """休暇申請を却下"""
        return self.client._request(
            "POST",
            f"/leaves/requests/{request_id}/reject",
            data={"reason": reason}
        )


class ExpensesAPI:
    """経費管理API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def submit(
        self,
        employee_id: str,
        amount: float,
        category: str,
        date: str,
        description: str,
        receipt_id: Optional[str] = None,
        project_code: Optional[str] = None
    ) -> Expense:
        """経費申請"""
        data = {
            "employee_id": employee_id,
            "amount": amount,
            "category": category,
            "date": date,
            "description": description
        }
        if receipt_id:
            data["receipt_id"] = receipt_id
        if project_code:
            data["project_code"] = project_code
        
        result = self.client._request("POST", "/expenses", data=data)
        return Expense(**result)
    
    def upload_receipt(
        self,
        file_path: str,
        employee_id: str,
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """レシートをアップロード"""
        with open(file_path, "rb") as f:
            files = {"receipt": f}
            data = {"employee_id": employee_id}
            if category:
                data["category"] = category
            
            return self.client._request(
                "POST",
                "/expenses/receipts",
                data=data,
                files=files
            )
    
    def list(
        self,
        employee_id: Optional[str] = None,
        status: Optional[str] = None,
        category: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 20,
        offset: int = 0
    ) -> Dict[str, Any]:
        """経費申請一覧を取得"""
        params = {
            "limit": limit,
            "offset": offset
        }
        if employee_id:
            params["employee_id"] = employee_id
        if status:
            params["status"] = status
        if category:
            params["category"] = category
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        
        return self.client._request("GET", "/expenses", params=params)
    
    def approve(self, expense_id: str) -> Expense:
        """経費申請を承認"""
        data = self.client._request("POST", f"/expenses/{expense_id}/approve")
        return Expense(**data)


class ReportsAPI:
    """レポート・分析API"""
    
    def __init__(self, client: AIOSClient):
        self.client = client
    
    def generate(self, report_type: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """レポートを生成"""
        return self.client._request(
            "POST",
            f"/reports/{report_type}",
            data=params or {}
        )
    
    def get_dashboard(self, dashboard_id: str) -> Dict[str, Any]:
        """ダッシュボードデータを取得"""
        return self.client._request("GET", f"/reports/dashboards/{dashboard_id}")
    
    def get_human_capital_metrics(
        self,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        group_by: Optional[str] = None
    ) -> Dict[str, Any]:
        """人的資本指標を取得"""
        params = {}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        if group_by:
            params["group_by"] = group_by
        
        return self.client._request("GET", "/reports/human-capital", params=params)

File: sdk/typescript/ai-os-sdk.ts
/**
 * AI-OS TypeScript SDK
 * @version 1.0.0
 * @description AI-OSプラットフォームとの統合を簡単にするためのTypeScript SDK
 */

export interface AIOSConfig {
  apiKey: string;
  baseUrl?: string;
  version?: string;
  timeout?: number;
  retryAttempts?: number;
  retryDelay?: number;
}

export interface Employee {
  id: string;
  employeeCode: string;
  name: string;
  email: string;
  department: string;
  position: string;
  hireDate: string;
  accountStatus: 'active' | 'inactive' | 'suspended';
  metadata?: Record<string, any>;
}

export interface TimeRecord {
  id: string;
  employeeId: string;
  type: 'clock_in' | 'clock_out' | 'break_start' | 'break_end';
  timestamp: string;
  location?: {
    latitude: number;
    longitude: number;
  };
  device?: string;
}

export interface PayrollResult {
  batchId: string;
  yearMonth: string;
  employeeCount: number;
  totalAmount: number;
  status: 'processing' | 'completed' | 'approved' | 'failed';
  createdAt: string;
  completedAt?: string;
}

export interface Payslip {
  id: string;
  employeeId: string;
  yearMonth: string;
  baseSalary: number;
  overtimePay: number;
  allowances: number;
  grossPay: number;
  socialInsurance: number;
  incomeTax: number;
  totalDeductions: number;
  netPay: number;
}

export interface LeaveRequest {
  id: string;
  employeeId: string;
  type: 'paid' | 'sick' | 'special' | 'unpaid';
  startDate: string;
  endDate: string;
  reason: string;
  status: 'pending' | 'approved' | 'rejected' | 'cancelled';
  approvedBy?: string;
  approvedAt?: string;
}

export interface LeaveBalance {
  employeeId: string;
  paidLeave: {
    total: number;
    used: number;
    remaining: number;
    expiringWithin90Days: number;
    nextGrantDate: string;
  };
  sickLeave: {
    remaining: number;
  };
}

export interface Expense {
  id: string;
  employeeId: string;
  amount: number;
  category: string;
  date: string;
  description: string;
  status: 'pending' | 'approved' | 'rejected' | 'reimbursed';
  receiptId?: string;
  projectCode?: string;
  aiRecommendation?: 'approve' | 'review' | 'reject';
}

export interface HumanCapitalMetrics {
  employeeCount: number;
  averageTenure: number;
  turnoverRate: number;
  engagementScore: number;
  productivityIndex: number;
  departments: Array<{
    id: string;
    name: string;
    turnoverRisk: number;
  }>;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

export interface ErrorResponse {
  message: string;
  code?: string;
  details?: any;
}

export class AIOSError extends Error {
  constructor(
    message: string,
    public statusCode: number,
    public response?: ErrorResponse
  ) {
    super(message);
    this.name = 'AIOSError';
  }
}

export class AIOSClient {
  private apiKey: string;
  private baseUrl: string;
  private version: string;
  private timeout: number;
  private retryAttempts: number;
  private retryDelay: number;

  constructor(config: AIOSConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl || 'https://api.ai-os.com';
    this.version = config.version || 'v1';
    this.timeout = config.timeout || 30000;
    this.retryAttempts = config.retryAttempts || 3;
    this.retryDelay = config.retryDelay || 1000;

    if (!this.apiKey) {
      throw new Error('APIキーが必要です。configで設定してください。');
    }
  }

  /**
   * HTTPリクエストを実行
   */
  private async request<T>(
    method: string,
    endpoint: string,
    data?: any,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${this.baseUrl}/api/${this.version}${endpoint}`;
    const headers: HeadersInit = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'X-SDK-Version': '1.0.0',
      'X-SDK-Language': 'TypeScript',
      ...options.headers
    };

    const config: RequestInit = {
      method,
      headers,
      signal: AbortSignal.timeout(this.timeout),
      ...options
    };

    if (data && ['POST', 'PUT', 'PATCH'].includes(method)) {
      config.body = JSON.stringify(data);
    }

    let lastError: Error | undefined;
    
    for (let attempt = 0; attempt < this.retryAttempts; attempt++) {
      try {
        const response = await fetch(url, config);
        
        if (!response.ok) {
          const error: ErrorResponse = await response.json();
          throw new AIOSError(
            error.message || 'APIエラー',
            response.status,
            error
          );
        }
        
        return await response.json();
      } catch (error) {
        lastError = error as Error;
        if (attempt < this.retryAttempts - 1) {
          await this.sleep(this.retryDelay * Math.pow(2, attempt));
        }
      }
    }
    
    throw lastError;
  }

  /**
   * スリープユーティリティ
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * 従業員管理API
   */
  public readonly employees = {
    list: async (params?: {
      limit?: number;
      offset?: number;
      department?: string;
      status?: string;
    }): Promise<PaginatedResponse<Employee>> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<PaginatedResponse<Employee>>(
        'GET',
        `/employees${queryString ? '?' + queryString : ''}`
      );
    },

    get: async (employeeId: string): Promise<Employee> => {
      return this.request<Employee>('GET', `/employees/${employeeId}`);
    },

    create: async (data: Omit<Employee, 'id'>): Promise<Employee> => {
      return this.request<Employee>('POST', '/employees', data);
    },

    update: async (
      employeeId: string,
      data: Partial<Employee>
    ): Promise<Employee> => {
      return this.request<Employee>('PUT', `/employees/${employeeId}`, data);
    },

    delete: async (employeeId: string): Promise<void> => {
      return this.request<void>('DELETE', `/employees/${employeeId}`);
    }
  };

  /**
   * 勤怠管理API
   */
  public readonly timeRecords = {
    clock: async (data: {
      employeeId: string;
      type: TimeRecord['type'];
      timestamp?: string;
      location?: TimeRecord['location'];
    }): Promise<TimeRecord> => {
      return this.request<TimeRecord>('POST', '/time-records/clock', data);
    },

    clockIn: async (employeeId: string): Promise<TimeRecord> => {
      return this.request<TimeRecord>('POST', '/time-records/clock-in', { employeeId });
    },

    clockOut: async (employeeId: string): Promise<TimeRecord> => {
      return this.request<TimeRecord>('POST', '/time-records/clock-out', { employeeId });
    },

    list: async (params?: {
      employeeId?: string;
      startDate?: string;
      endDate?: string;
      limit?: number;
      offset?: number;
    }): Promise<PaginatedResponse<TimeRecord>> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<PaginatedResponse<TimeRecord>>(
        'GET',
        `/time-records${queryString ? '?' + queryString : ''}`
      );
    },

    getMonthlySummary: async (
      employeeId: string,
      yearMonth: string
    ): Promise<{
      totalHours: number;
      overtimeHours: number;
      holidayWork: number;
      nightWork: number;
      absences: number;
    }> => {
      return this.request(
        'GET',
        `/time-records/summary/${employeeId}/${yearMonth}`
      );
    }
  };

  /**
   * 給与管理API
   */
  public readonly payroll = {
    calculate: async (data: {
      yearMonth: string;
      departmentIds?: string[];
      employeeIds?: string[];
      includeBonus?: boolean;
      calculateTax?: boolean;
      calculateInsurance?: boolean;
    }): Promise<PayrollResult> => {
      return this.request<PayrollResult>('POST', '/payroll/calculate', data);
    },

    getPayslip: async (
      employeeId: string,
      yearMonth: string
    ): Promise<Payslip> => {
      return this.request<Payslip>(
        'GET',
        `/payroll/payslips/${employeeId}/${yearMonth}`
      );
    },

    listPayslips: async (params?: {
      yearMonth?: string;
      departmentId?: string;
      limit?: number;
      offset?: number;
    }): Promise<PaginatedResponse<Payslip>> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<PaginatedResponse<Payslip>>(
        'GET',
        `/payroll/payslips${queryString ? '?' + queryString : ''}`
      );
    },

    approve: async (batchId: string): Promise<{
      status: string;
      approvedAt: string;
      approvedBy: string;
    }> => {
      return this.request('POST', `/payroll/batches/${batchId}/approve`);
    }
  };

  /**
   * 休暇管理API
   */
  public readonly leaves = {
    request: async (data: {
      employeeId: string;
      type: LeaveRequest['type'];
      startDate: string;
      endDate: string;
      reason: string;
    }): Promise<LeaveRequest> => {
      return this.request<LeaveRequest>('POST', '/leaves/requests', data);
    },

    listRequests: async (params?: {
      employeeId?: string;
      status?: string;
      startDate?: string;
      endDate?: string;
      limit?: number;
      offset?: number;
    }): Promise<PaginatedResponse<LeaveRequest>> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<PaginatedResponse<LeaveRequest>>(
        'GET',
        `/leaves/requests${queryString ? '?' + queryString : ''}`
      );
    },

    getBalance: async (employeeId: string): Promise<LeaveBalance> => {
      return this.request<LeaveBalance>('GET', `/leaves/balance/${employeeId}`);
    },

    approve: async (requestId: string): Promise<LeaveRequest> => {
      return this.request<LeaveRequest>(
        'POST',
        `/leaves/requests/${requestId}/approve`
      );
    },

    reject: async (
      requestId: string,
      reason: string
    ): Promise<LeaveRequest> => {
      return this.request<LeaveRequest>(
        'POST',
        `/leaves/requests/${requestId}/reject`,
        { reason }
      );
    }
  };

  /**
   * 経費管理API
   */
  public readonly expenses = {
    submit: async (data: {
      employeeId: string;
      amount: number;
      category: string;
      date: string;
      description: string;
      receiptId?: string;
      projectCode?: string;
    }): Promise<Expense> => {
      return this.request<Expense>('POST', '/expenses', data);
    },

    uploadReceipt: async (
      file: File,
      metadata?: Record<string, string>
    ): Promise<{
      receiptId: string;
      merchantName: string;
      amount: number;
      date: string;
      confidence: number;
    }> => {
      const formData = new FormData();
      formData.append('receipt', file);
      
      if (metadata) {
        Object.entries(metadata).forEach(([key, value]) => {
          formData.append(key, value);
        });
      }

      return this.request('POST', '/expenses/receipts', null, {
        headers: { 'Content-Type': undefined as any },
        body: formData
      });
    },

    list: async (params?: {
      employeeId?: string;
      status?: string;
      category?: string;
      startDate?: string;
      endDate?: string;
      limit?: number;
      offset?: number;
    }): Promise<PaginatedResponse<Expense>> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<PaginatedResponse<Expense>>(
        'GET',
        `/expenses${queryString ? '?' + queryString : ''}`
      );
    },

    approve: async (expenseId: string): Promise<Expense> => {
      return this.request<Expense>('POST', `/expenses/${expenseId}/approve`);
    }
  };

  /**
   * レポート・分析API
   */
  public readonly reports = {
    generate: async (
      reportType: string,
      params?: Record<string, any>
    ): Promise<{
      reportId: string;
      status: string;
      downloadUrl?: string;
    }> => {
      return this.request('POST', `/reports/${reportType}`, params);
    },

    getDashboard: async (dashboardId: string): Promise<{
      currentlyWorking: number;
      overtimePrediction: number;
      laborCostProgress: number;
      alerts: Array<{
        type: string;
        severity: string;
        message: string;
      }>;
    }> => {
      return this.request('GET', `/reports/dashboards/${dashboardId}`);
    },

    getHumanCapitalMetrics: async (params?: {
      startDate?: string;
      endDate?: string;
      groupBy?: string;
    }): Promise<HumanCapitalMetrics> => {
      const queryString = params ? new URLSearchParams(params as any).toString() : '';
      return this.request<HumanCapitalMetrics>(
        'GET',
        `/reports/human-capital${queryString ? '?' + queryString : ''}`
      );
    }
  };
}